### PR TITLE
Switch attribute name from PEP658 'data-dist-info-metadata' to PEP714 'data-core-metadata'

### LINKED ISSUE #103: Support PEP 714
https://peps.python.org/pep-0714/

### PR DESCRIPTION
**Pull request by derek-keeler**
Fixes #103

### CODE REVIEW

**Comment by derek-keeler** [comment]
Please ensure my decision to `raise ValueError(...)` was correct. I took `MUST` as such.

File: mousebender/simple.py [hunk 1]
```diff
@@ -263,8 +263,27 @@ def handle_starttag(
...
+            #   representation MAY also be emitted using the data-dist-info-metadata, and
+            #   if it does so it MUST match the value of data-core-metadata.
+            if ("data-dist-info-metadata" in attrs) and (
+                found_metadata != attrs.get("data-dist-info-metadata")
+            ):
+                # Because of the use of MUST we will throw an exception in this case.
+                raise ValueError(
+                    "PEP 714 Violation. `data-dist-info-metadata` is present in package ",
+                    "metadata, but it does not match the value given in `data-core-metadata`",
+                )
```
Comment (brettcannon) [inline mousebender/simple.py:285]:
```suggestion
                raise ValueError(
                    "'data-dist-info-metadata' differs from 'data-core-metadata'",
                )
```

We should probably raise a new `InvalidDistroMetadata` exception.

File: mousebender/simple.py
Comment (brettcannon) [inline mousebender/simple.py:1]: Missing updates to the the appropriate `TypedDict`.

File: tests/test_simple.py [hunk 2]
```diff
@@ -390,27 +390,94 @@ def test_default(self):
...
         ],
     )
-    def test_hash(self, attribute: str):
+    def test_ignore_pep_658_dist_info_metadata_attr_name(self, attribute: str):
         html = f'<a href="spam-1.2.3-py3.none.any.whl" {attribute}>spam-1.2.3-py3.none.any.whl</a>'
         details = simple.from_project_details_html(html, "test_default")
         assert len(details["files"]) == 1
+        assert "dist-info-metadata" not in details["files"][0]
+
+    # PEP-714 replaces attribute named 'data-dist-info-metadata' with data-core-metadata
```
Comment (brettcannon) [inline tests/test_simple.py:430]:
```suggestion
    # PEP 714 replaces data-dist-info-metadata with data-core-metadata
```

File: tests/test_simple.py [hunk 3]
```diff
@@ -390,27 +390,94 @@ def test_default(self):
...
+    # while still allowing for it's presence. The original named attribute being present
```
Comment (brettcannon) [inline tests/test_simple.py:431]:
```suggestion
    # while still allowing for its presence. The original named attribute being present
```

File: tests/test_simple.py [hunk 4]
```diff
@@ -390,27 +390,94 @@ def test_default(self):
...
+    )
+    def test_hash_with_both_pep_658_and_pep_714_metadata_attr_names_matching_values(
+        self, attr_pep_658: str, attr_pep_714: str
+    ):
+        html = f'<a href="spam-1.2.3-py3.none.any.whl" {attr_pep_658} {attr_pep_714}>spam-1.2.3-py3.none.any.whl</a>'
+        details = simple.from_project_details_html(html, "test_default")
+        assert len(details["files"]) == 1
         assert details["files"][0]["dist-info-metadata"] == {"sha256": "abcdef"}
 
+    # PEP-714 replaces attribute named 'data-dist-info-metadata' with data-core-metadata
```
Comment (brettcannon) [inline tests/test_simple.py:456]:
```suggestion
    # PEP 714 replaces data-dist-info-metadata with data-core-metadata
```

File: tests/test_simple.py [hunk 5]
```diff
@@ -390,27 +390,94 @@ def test_default(self):
...
+    # while still allowing for it's presence. The original named attribute being present
```
Comment (brettcannon) [inline tests/test_simple.py:457]:
```suggestion
    # while still allowing for its presence. The original named attribute being present
```

File: tests/test_simple.py
Comment (brettcannon) [inline tests/test_simple.py:1]: Should add a test for when `data-dist-info-metadata` is just defined in the HTML without a hash, both by itself and when `data-core-metadata` is also set.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Thanks for working on this! I think there are some test cases to cover and a missing update to one of the `TypedDict`s.

**Comment by brettcannon** [comment]
Also forgot to mention the docs will need updating (adding the PEP to the list of supported PEPs, the new field as necessary, etc.).

File: tests/test_simple.py (hunk 2 above)
Comment (derek-keeler) [reply tests/test_simple.py:430]: Replaced each instance of this one...

File: tests/test_simple.py (hunk 3 above)
Comment (derek-keeler) [reply tests/test_simple.py]: English better, me.

File: tests/test_simple.py
Comment (derek-keeler) [reply tests/test_simple.py:1]: Are we concerned about the values being different IF one is `true` and the other is `TRUE`? The attribute _is_ technically set differently and could be argued as such, but in the special case of `True` do we want to accept case insensitive comparisons? (See the last test case in `test_hash_with_both_pep_658_and_pep_714_metadata_attr_names_unmatched_values` for what I'm going on about).

File: mousebender/simple.py
Comment (derek-keeler) [reply mousebender/simple.py:1]: I think I've got it. Also I believe we are updating the "_meta" version. However, I'm unclear on this:

Do we _remove_ the `dist-info-metadata` from the `_OptionalProjectFileDetails_1_2`, or leave it there (as some clients _MAY_ choose to read it, if it is there and `core-metadata` is not)?

Comment (derek-keeler) [reply mousebender/simple.py:1]: Ha! Answered my own question.  They are a union, and so removal is not really an option.

Comment (derek-keeler) [reply mousebender/simple.py:1]: Alright, I'm done thinking about this for a bit. I think I'll need a hand with versioning this properly - do we need an `ACCEPT_JSON_V2` now? Seems like it...?

File: tests/test_simple.py
Comment (brettcannon) [reply tests/test_simple.py:1]: `TRUE` is not a valid value according to https://peps.python.org/pep-0658/#specification .

File: mousebender/simple.py
Comment (brettcannon) [reply mousebender/simple.py:1]: The meta version was **not** bumped by PEP 714.

And don't remove the old key as it's still valid if a server returns it (especially for backwards-compatibility).

Comment (derek-keeler) [reply mousebender/simple.py:1]: That's great, thank you. I felt I was going down a bit of a rabbit hole there. (It _was_ getting past my bedtime haha).

File: tests/test_simple.py
Comment (derek-keeler) [reply tests/test_simple.py:1]: Beauty, solved.

**Pull request closed without merging**